field_audit/models.py

```diff
@@ -0,0 +1,342 @@
+        init_values = cls.reset_initial_values(field_names, instance)
+        if is_create or is_delete:
+            # initial values are meaningless in these scenarios, discard them
+            init_values = {}
+
+        event = None
+        changes = []
+        for field_name in field_names:
+            kwargs = {
+                "field_name": field_name,
+                "value": cls.get_field_value(instance, field_name),
+                "is_create": is_create,
+                "is_delete": is_delete,
+            }
+            if init_values:
+                try:
+                    kwargs["init_value"] = init_values[field_name]
+                except KeyError:
+                    pass
+            change = FieldChange.create_if_changed(**kwargs)
+            if change is not None:
+                # only instantiate the top-level event once we know we need it
+                event = lazy_event()
+                change.event = event
+                changes.append(change)
+        if changes:
+            with transaction.atomic():
+                event.save()
+                FieldChange.objects.bulk_create(changes)
+
+    def __repr__(self):  # pragma: no cover
+        cls_name = type(self).__name__
+        return f"<{cls_name} ({self.id}, {self.object_class_path!r})>"
+
+
+class FieldChange(models.Model):
+
+    event = models.ForeignKey(
+        AuditEvent,
+        on_delete=models.CASCADE,
+        related_name="changes",
+    )
+    field_name = models.CharField(db_index=True, max_length=127)
+    delta = models.JSONField()
+
+    objects = get_manager("FIELDCHANGE_MANAGER", models.Manager)
+
+    class MISSING:
+        pass
+    MISSING = MISSING()  # a unique object that reprs nicely
+
+    class Meta:
+        unique_together = [("event", "field_name")]
+
+    @classmethod
+    def create_if_changed(cls, field_name, value, is_create, is_delete,
+                          init_value=MISSING):
+        """Factory method to create a new ``FieldChange`` containing the details
+        of a field change on a Model object.
+
+        :param field_name: name of the field that may have changed
+        :param value: current value of the field (i.e. value present when the DB
+            write (e.g. ``save()``, ``delete()``, etc) operation occurs.
+        :param is_create: whether or not the change is creating ``instance``
+        :param is_delete: whether or not the change is deleting ``instance``
+        :param init_value: value of the field before it was changed (not used
+            when either ``is_delete`` or ``is_create`` are True, providing a
+            value in either scenario will raise an exception)
+        :returns: new instance of a ``FieldChange`` object or ``None``
+        :raises: ``ValueError``
+        """
+        missing = cls.MISSING  # create a local reference
+        if value is missing:
+            raise ValueError("'value' cannot be MISSING")
```

Comment:
In this scenario the behavior varies (depending on other args), which is why I've explicitly prohibited it. If you pass `FieldChange.MISSING` when either `(is_create or is_delete) == True`, then the eventual "change check" [here](https://github.com/dimagi/django-field-audit/blob/4808788a04467879e8da546895c8dbcc3b732d3c/field_audit/models.py#L331) (`if old_value == new_value:`) will be `False`, resulting in a "calculated no change" scenario which will not create a new `AuditEvent` record. This would be an error condition because `is_create` _or_ `is_delete` _always_ implies there has been a change. Perhaps this function could be refactored to handled this in a more elegant way?

This could be made more efficient (which would result in the JSON serializer to raising a `ValueError` later on if you pass `value=FieldChange.MISSING`):

```diff
-        if value is missing:
-            raise ValueError("'value' cannot be MISSING")
-        if (is_create or is_delete) and init_value is not missing:
-            raise ValueError(
-                "'init_value' is mutually exclusive with 'is_create' and "
-                "'is_delete'"
-            )
+        if is_create or is_delete:
+            if value is missing:
+                raise ValueError("'value' cannot be MISSING")
+            if init_value is not missing:
+                raise ValueError(
+                    "'init_value' is mutually exclusive with 'is_create' and "
+                    "'is_delete'"
+                )
```

Reply:
Changed in 7898b0a.